File: Discrete Math/sorting algorithms/selection_sort.py

def selection_sort(arr):
    n = len(arr)
    
    for i in range(n - 1):
        min_idx = i
        
        for j in range(i + 1, n):
            if arr[j] < arr[min_idx]:
                min_idx = j
                        
        arr[i], arr[min_idx] = arr[min_idx], arr[i]
        
    return arr


arr = [64, 25, 12, 22, 11]
print(selection_sort(arr))  # Output: [11, 12, 22, 25, 64]

File: dsa/chapter_5_sort/selection_sort.cpp
#include <iostream>
#include <vector>
#include <algorithm>

void selectionSort(std::vector<int>& arr) {
    int n = arr.size();
    for (int i = 0; i < n; ++i) {
        int min_index = i;
        for (int j = i + 1; j < n;
        ++j) {
            if (arr[j] < arr[min_index]) {
                min_index = j;
            }
        }
        
        if (min_index != i) {
            std::swap(arr[i], arr[min_index]);
        }
    }
}

// Helper function to print the array
void printArray(const std::vector<int>& arr) {
    for (int val : arr) {
        std::cout << val << " ";
    }
    std::cout << std::endl;
}

int main() {
    std::vector<int> my_list = {6, 4, 1, 5, 3, 2};

    std::cout << "Original list: ";
    printArray(my_list);

    selectionSort(my_list);

    std::cout << "Sorted list: ";
    printArray(my_list);

    return 0;
}